test/sql/test_temporal.py

```diff
@@ -0,0 +1,356 @@
+"""Tests for temporal table structure, including system and
+application versioning"""
+from sqlalchemy import Column
+from sqlalchemy import Integer
+from sqlalchemy import MetaData
+from sqlalchemy import Period
+from sqlalchemy import PrimaryKeyConstraint
+from sqlalchemy import schema
+from sqlalchemy import SystemTimePeriod
+from sqlalchemy import Table
+from sqlalchemy import testing
+from sqlalchemy import UniqueConstraint
+from sqlalchemy.exc import ArgumentError
+from sqlalchemy.sql.sqltypes import DATE
+from sqlalchemy.sql.sqltypes import TIMESTAMP
+from sqlalchemy.sql.sqltypes import VARCHAR
+from sqlalchemy.testing import assert_raises_message
+from sqlalchemy.testing import AssertsCompiledSQL
+from sqlalchemy.testing import eq_
+from sqlalchemy.testing import fixtures
+from sqlalchemy.testing import is_
+from sqlalchemy.testing import is_not_none
+
+
+class PeriodTest(fixtures.TestBase, AssertsCompiledSQL):
+    """Test the basic period construct"""
+
+    __dialect__ = "default"
+
+    @testing.combinations(lambda x: x, lambda x: x.name)
+    def test_period(self, fn):
+        """Test basic period construct specifying object and str columns"""
+        m = MetaData()
+        start_ts = Column("start_ts", TIMESTAMP)
+        end_ts = Column("end_ts", TIMESTAMP)
+        t = Table(
+            "t",
+            m,
+            start_ts,
+            end_ts,
+            Period("test_period", fn(start_ts), fn(end_ts)),
+        )
+        self.assert_compile(
+            schema.CreateTable(t),
+            "CREATE TABLE t ("
+            "start_ts TIMESTAMP, "
+            "end_ts TIMESTAMP, "
+            "PERIOD FOR test_period (start_ts, end_ts))",
+        )
+
+    def test_period_pk_col_arg(self):
+        """Test setting a primary key on a PERIOD via column/period args"""
+        m = MetaData()
+        t = Table(
+            "t",
+            m,
+            Column("id", Integer, primary_key=True),
+            Column("start_ts", TIMESTAMP),
+            Column("end_ts", TIMESTAMP),
+            Period("test_period", "start_ts", "end_ts", primary_key=True),
+        )
+        self.assert_compile(
+            schema.CreateTable(t),
+            "CREATE TABLE t ("
+            "id INTEGER NOT NULL, "
+            "start_ts TIMESTAMP, "
+            "end_ts TIMESTAMP, "
+            "PERIOD FOR test_period (start_ts, end_ts), "
+            "PRIMARY KEY (id, test_period WITHOUT OVERLAPS))",
```

Comment:
**mike bayer** (zzzeek) wrote:

wow, ok, this is another new syntax for me, this has to be worked out.  can there be more than one period in a PK ?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3803

Reply:
**Trevor Gross** (tgross35) wrote:

I mentioned this in a comment elsewhere but the spec currently only allows one AV period per table, so not really at this point.

From my testing on Maria, the period always has to be the last thing in the PK directly followed by `WITHOUT OVERLAPS` (the extra keyword isn't optional). So I don't really know if `WITHOUT OVERLAPS` stays with the period, or if it is always at the end of the PK.

It might be spelled out better in the ISO spec if you have it and are able to take a peek. However, as-is the generated SSL works on the only backend with application versioning and matches the examples in the temporal doc, so really probably only worth crossing that bridge after it's built.

Edit: looks like IBM actually has this implemented too, and the syntax matches https://www.ibm.com/docs/en/db2/11.5?topic=aptt-creating

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3803